cpp/src/streaming/cudf/table_chunk.cpp

```diff
@@ -0,0 +1,151 @@
+    data_alloc_size_[static_cast<std::size_t>(MemoryType::DEVICE)] = table_->alloc_size();
+    make_available_cost_ = 0;
+}
+
+TableChunk::TableChunk(
+    std::uint64_t sequence_number,
+    cudf::table_view table_view,
+    std::size_t device_alloc_size,
+    rmm::cuda_stream_view stream
+)
+    : sequence_number_{sequence_number}, table_view_{table_view}, stream_{stream} {
+    data_alloc_size_[static_cast<std::size_t>(MemoryType::DEVICE)] = device_alloc_size;
+    make_available_cost_ = 0;
+}
+
+TableChunk::TableChunk(
+    std::uint64_t sequence_number,
+    std::unique_ptr<cudf::packed_columns> packed_columns,
+    rmm::cuda_stream_view stream
+)
+    : sequence_number_{sequence_number},
+      packed_columns_{std::move(packed_columns)},
+      stream_{stream} {
+    RAPIDSMPF_EXPECTS(
+        packed_columns_ != nullptr,
+        "packed columns pointer cannot be null",
+        std::invalid_argument
+    );
+    table_view_ = cudf::unpack(*packed_columns_);
+    data_alloc_size_[static_cast<std::size_t>(MemoryType::DEVICE)] =
+        packed_columns_->gpu_data->size();
+    make_available_cost_ = 0;
+}
+
+TableChunk::TableChunk(
+    std::uint64_t sequence_number,
+    std::unique_ptr<PackedData> packed_data,
+    rmm::cuda_stream_view stream
+)
+    : sequence_number_{sequence_number},
+      packed_data_{std::move(packed_data)},
+      stream_{stream} {
+    RAPIDSMPF_EXPECTS(
+        packed_data_ != nullptr,
+        "packed data pointer cannot be null",
+        std::invalid_argument
+    );
+    RAPIDSMPF_EXPECTS(
+        !packed_data_->empty(), "packed data cannot be empty", std::invalid_argument
+    );
+    data_alloc_size_[static_cast<std::size_t>(packed_data_->data->mem_type())] =
+        packed_data_->data->size;
+    if (packed_data_->data->mem_type() == MemoryType::HOST) {
+        make_available_cost_ = packed_data_->data->size;
+    } else {
+        make_available_cost_ = 0;
+    }
+}
+
+std::uint64_t TableChunk::sequence_number() const noexcept {
+    return sequence_number_;
+}
+
+std::size_t TableChunk::data_alloc_size(MemoryType mem_type) const {
+    return data_alloc_size_.at(static_cast<std::size_t>(mem_type));
+}
+
+bool TableChunk::is_available() const noexcept {
+    return table_view_.has_value();
+}
+
+std::size_t TableChunk::make_available_cost() const noexcept {
+    return make_available_cost_;
+}
+
+TableChunk TableChunk::make_available(
+    MemoryReservation& reservation, rmm::cuda_stream_view stream, BufferResource* br
+) {
+    if (is_available()) {
+        return std::move(*this);
+    }
+    RAPIDSMPF_EXPECTS(packed_data_ != nullptr, "packed data pointer cannot be null");
+    PackedData packed_data = std::move(*packed_data_);
+    return TableChunk{
+        sequence_number_,
+        std::make_unique<cudf::packed_columns>(
+            std::move(packed_data.metadata),
+            br->move_to_device_buffer(std::move(packed_data.data), stream, reservation)
+        ),
+        stream
+    };
+}
```

Comment:
*issue*: We should use the buffer resource of the _reservation_ (and not require a separate `BufferResource` here).

Reply:
good point, done